src/Compilers/CSharp/Portable/Symbols/TypeWithState.cs

```diff
@@ -87,7 +87,7 @@ public TypeWithAnnotations ToTypeWithAnnotations(CSharpCompilation compilation,
             if (Type?.IsTypeParameterDisallowingAnnotationInCSharp8() == true)
             {
                 var type = TypeWithAnnotations.Create(Type, NullableAnnotation.NotAnnotated);
-                return State == NullableFlowState.MaybeDefault ? type.SetIsAnnotated(compilation) : type;
+                return (State == NullableFlowState.MaybeDefault || asAnnotatedType) ? type.SetIsAnnotated(compilation) : type;
```

Comment:
Was this all that was needed to fix the 'foreach' scenario?

Reply:
Yes, the `for` loop starts with a local declaration. That's handled by `VisitLocalDeclaration`, which correctly calls `ToAnnotatedTypeWithAnnotations` when the local has an inferred type, but when we get to `ToTypeWithAnnotations` the annotation would get dropped.